packages/mattermost-redux/src/utils/apps.ts

```diff
@@ -1,116 +1,188 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import {AppBinding} from '../types/apps';
-import {AppBindingLocations} from '../constants/apps';
-
-// fillAndTrimBindingsInformation does:
-// - Build the location (e.g. channel_header/binding)
-// - Inherit app calls
-// - Inherit app ids
-// - Trim invalid bindings (do not have an app call or app id at the leaf)
-export function fillAndTrimBindingsInformation(binding?: AppBinding) {
+import {AppBinding, AppField, AppForm} from '../types/apps';
+import {AppBindingLocations, AppFieldTypes} from '../constants/apps';
+
+export function cleanBinding(binding: AppBinding, topLocation: string) {
     if (!binding) {
         return;
     }
 
-    binding.bindings?.forEach((b) => {
-        // Propagate id down if not defined
+    const toRemove: number[] = [];
+    const usedLabels: {[label: string]: boolean} = {};
+    binding.bindings?.forEach((b, i) => {
+        // Inheritance and defaults
+        if (!b.call) {
+            b.call = binding.call;
+        }
+
+        if (b.form) {
+            cleanForm(b.form);
+        } else {
+            b.form = binding.form;
+        }
+
         if (!b.app_id) {
             b.app_id = binding.app_id;
         }
 
-        // Compose location
+        if (!b.label) {
+            b.label = b.location || '';
+        }
+
         b.location = binding.location + '/' + b.location;
 
-        // Propagate call down if not defined
-        if (!b.call) {
-            b.call = binding.call;
+        // Validation
+        if (!b.label) {
+            toRemove.unshift(i);
+            return;
         }
 
-        fillAndTrimBindingsInformation(b);
-    });
-
-    // Trim branches without app_id
-    if (!binding.app_id) {
-        binding.bindings = binding.bindings?.filter((v) => v.app_id);
-    }
+        switch (topLocation) {
+        case AppBindingLocations.COMMAND: {
+            if (b.label.match(/ |\t/)) {
+                toRemove.unshift(i);
+                return;
+            }
+
+            if (usedLabels[b.label]) {
+                toRemove.unshift(i);
+                return;
+            }
+            break;
+        }
+        case AppBindingLocations.CHANNEL_HEADER_ICON: {
+            if (!b.icon) {
+                toRemove.unshift(i);
+                return;
+            }
+            break;
```

Comment:
Is this checking also server side?

Reply:
This is a "webapp only problem". In mobile we don't need the image, because it is always a menu.

Reply:
That being said, we can check the context, and filter those on server side if we are serving the webapp. Will do in my server side PR.